sound/soc/sof/topology.c

```diff
@@ -1256,15 +1256,43 @@ static int sof_connect_dai_widget(struct snd_soc_component *scomp,
 
 		switch (w->id) {
 		case snd_soc_dapm_dai_out:
-			for_each_rtd_cpu_dai(rtd, i, cpu_dai)
-				cpu_dai->capture_widget = w;
+			for_each_rtd_cpu_dai(rtd, i, cpu_dai) {
+				/*
+				 * The order of DAI widgets defined in topology
+				 * doesnt matter as all the CPU DAI's will be
+				 * enabled/disabled simultaneously
+				 */
+				if (!cpu_dai->capture_widget)
+					break;
+			}
+			if (cpu_dai->capture_widget) {
```

Comment:
This isn't safe. If you didn't find a free slot i.e. if you didn't break out of the above loop and it ran until its termination, it means the stop condition of the loop was triggered, which can mean, that cpu_dai == NULL. In general it isn't safe to dereference the iterator after a "list_for_each()" style loop has possibly terminated normally. This isn't a list-traversing loop, but it still doesn't look safe.